repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/PreProcessorUtils.java

```diff
@@ -200,4 +221,177 @@ public static void verifyDuplicateAssetByName(String typeName, String assetName,
+        mustArray.add(mapOf("term", mapOf(LEXICOGRAPHICAL_SORT_ORDER, lexoRank)));
+        if(StringUtils.isNotEmpty(glossaryQualifiedName)) {
+            mustArray.add(mapOf("terms", mapOf("__typeName.keyword", Arrays.asList(ATLAS_GLOSSARY_TERM_TYPENAME, ATLAS_GLOSSARY_CATEGORY_TYPENAME))));
+            mustArray.add(mapOf("term", mapOf("__glossary", glossaryQualifiedName)));
+            String parentAttribute = isTerm ? "__categories" : "__parentCategory";
+            if(StringUtils.isEmpty(parentQualifiedName)) {
+                boolFilter.put("must_not", Arrays.asList(mapOf("exists", mapOf("field", parentAttribute))));
+            } else {
+                mustArray.add(mapOf("bool",mapOf("term", mapOf(parentAttribute, parentQualifiedName))));
+            }
+        } else{
+            mustArray.add(mapOf("terms", mapOf("__typeName.keyword", Arrays.asList(ATLAS_GLOSSARY_ENTITY_TYPE))));
+        }
+
+        boolFilter.put("must", mustArray);
+
+        return boolFilter;
+    }
+    public static Map<String, Object> generateDSLQueryForLastChild(String glossaryQualifiedName, String parentQualifiedName, boolean isTerm) {
+
+        Map<String, Object> sortKeyOrder = mapOf(LEXICOGRAPHICAL_SORT_ORDER, mapOf("order", "desc"));
+
+        Object[] sortArray = {sortKeyOrder};
+
+        Map<String, Object> boolMap = buildBoolQuery(glossaryQualifiedName, parentQualifiedName, isTerm);
+
+        Map<String, Object> dsl = new HashMap<>();
+        dsl.put("from", 0);
+        dsl.put("size", 1);
+        dsl.put("sort", sortArray);
+        dsl.put("query", mapOf("bool", boolMap));
+
+        return dsl;
+    }
+
+    private static Map<String, Object> buildBoolQuery(String glossaryQualifiedName, String parentQualifiedName, boolean isTerm) {
+        Map<String, Object> boolFilter = new HashMap<>();
+        List<Map<String, Object>> mustArray = new ArrayList<>();
+        List<Map<String, Object>> mustNotArray = new ArrayList<>();
+
+        mustArray.add(mapOf("term", mapOf("__state", "ACTIVE")));
+        if(StringUtils.isNotEmpty(glossaryQualifiedName)) {
+            String typeName = isTerm ? ATLAS_GLOSSARY_TERM_TYPENAME : ATLAS_GLOSSARY_CATEGORY_TYPENAME;
+            mustArray.add(mapOf("term", mapOf("__typeName.keyword", typeName)));
+            mustArray.add(mapOf("term", mapOf("__glossary", glossaryQualifiedName)));
+            String parentAttribute = isTerm ? "__categories" : "__parentCategory";
+            if(StringUtils.isEmpty(parentQualifiedName)) {
+                mustNotArray.add(mapOf("exists", mapOf("field", parentAttribute)));
+                boolFilter.put("must_not", mustNotArray);
+            }
+            else {
+                List<Map<String, Object>> shouldParentArray = new ArrayList<>();
+                shouldParentArray.add(mapOf("term", mapOf(parentAttribute, parentQualifiedName)));
+                mustArray.add(mapOf("bool",mapOf("should", shouldParentArray)));
```

Comment:
Why still using should here? this was already pointed out https://github.com/atlanhq/atlas-metastore/pull/3198#discussion_r1644169869